### ISSUE TITLE
git module: does not fetch new branches when branch points to the same commit

### ISSUE THREAD

**Issue by xytis** [ansible/ansible-modules-core#612]
##### Issue Type:

Bug Report
##### Ansible Version:

 a55c6625d4771c44017fce1d487b38749b12b381 (latest dev)
##### Environment:

Target OS: Ubuntu 14.04
Target git version 1.9.1
##### Summary:

Ansible git module does not fetch new branches when branch points to the same commit
##### Steps To Reproduce:

Create a simple git repository:

```
C1 (master)
```

Run such task:

```
- git: repo=<repo_url> version=master
```

Update the repo so that a new branch would point to the same commit:

```
C1 (master)(newbranch)
```

Rerun task:

```
- git: repo=<repo_url> version=newbranch
```
##### Expected Results:

Task updates the source.
##### Actual Results:

Task fails.

(moved from: https://github.com/ansible/ansible/issues/9876)

**Comment by xytis** [comment]
Still valid on https://github.com/ansible/ansible/commit/1a44d98916c110720bb040a274d9717ff7b14606

**Comment by xytis** [comment]
Related to https://github.com/ansible/ansible-modules-core/issues/527

**Comment by cjunge-work** [comment]
Just experienced the same error. Steps above basically the same as mine. Fixed it with a `git fetch origin` but it shouldn't be needed!

I commented on the related issue with the steps I had. #527

**Comment by bermudezjd** [comment]
I've also run into this issue.  Looking at the source code, it appears this problem was introduced in the 1.8.3 version here:
https://github.com/ansible/ansible-modules-core/blob/release1.8.3/source_control/git.py#L705-L706

The condition in 1.8.2 and before used to be:
https://github.com/ansible/ansible-modules-core/blob/release1.8.2/source_control/git.py#L687-L688

When the branches point to the same commit, the repo_updated flag is False.  In the prior versions, a switch would NOT have been performed.  IMHO, it would be preferable to switch over to the new branch, but for switch_version to work a fetch would have to be performed first.

The switch_version function attempts the following git command, which is where the failure occurs:

```
git checkout --force [version]
```

Failure is due to the fact that the new version in question has not been fetched.  I'm thinking that code could be simplified if a fetch is always done before attempting to checkout the version in question.

**Comment by FlorinAndrei** [comment]
Hit this very same problem today. The git module fails if I switch to a new branch which was just cloned from the old branch. Manually running "git fetch origin" fixes it. Nasty bug, took a while to track it down to Ansible.

**Comment by electrofelix** [comment]
Digging through the code suggests that the fix is to modify the lines at 
https://github.com/ansible/ansible-modules-core/blob/57c0b86db4af3f1de54509cecef5054f77466ae1/source_control/git.py#L760-L763 to also check if version is a remote branch and if it exists locally as well as checking if the tag is already local.

Adding the following method keeps the number of requests down to a single one:

```
def is_remote_branch_or_tag(git_path, module, dest, remote, version):
    cmd = ('%(git_path)s ls-remote %(remote)s -h refs/heads/%(version)s'
           '-t refs/tags/%(version)s' % (
               {'git_path': git_path, 'remote': remote, 'version': version}))
    (rc, out, err) = module.run_command(cmd, check_rc=True, cwd=dest)
    if version in out:
        return True
    else:
        return False
```

Then replace the above reference lines with:

```
            elif is_remote_branch_or_tag(git_path, module, dest, repo, version):
                # if the remote is a branch/tag and we have the it locally, exit early
                if version in get_tags(git_path, module, dest):
                    repo_updated = False
                if is_local_branch(git_path, module, dest, version):
                    repo_updated = False
```

This ensures that even if the requested branch has the same SHA1 as the current head (true if it's just been created) but has not been fetched yet, that a fetch will be performed. You could of course use the following as well without the need for another function to be added, but I'd prefer to avoid adding additional remote calls unless really necessary.

```
            elif is_remote_tag(git_path, module, dest, repo, version):
                # if the remote is a branch/tag and we have the it locally, exit early
                if version in get_tags(git_path, module, dest):
                    repo_updated = False
            elif is_remote_branch(git_path, module, dest, repo, version):
                if is_local_branch(git_path, module, dest, version):
                    repo_updated = False
```

**Comment by hansbogert** [comment]
I think I'm still experiencing this although somewhat different. Before I open a new issue with a slightly different reproduction I wanted to know if this is still an ongoing issue (after 1 year?!)

**Comment by electrofelix** [comment]
Looks like b77bf7a9aa1889fbee75f3db17c89816ca7c7838 contains a fix

**Comment by hansbogert** [comment]
@electrofelix how can b77bf7a be the fix if it predates the bug (or are you replying to someone else in this thread?)

**Comment by electrofelix** [comment]
@hansbogert sorry should have been clear. Different branches, AFAIK they aren't planning on backporting too many fixes to the 1.9 series, but it looks like it was fixed in the devel branch that was done for the 2.0 release. So depends on whether you are using 1.9 or 2.0.

**Comment by bhardin** [comment]
Any word on a fix here?

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/jctanner/ansibullbot/blob/ISSUE_TRIAGE_CLASS/ISSUE_HELP.md)

**Comment by jctanner** [comment]
My test playbook ...

```
[jtanner@fedmac AMC-612]$ cat site.yml 
- hosts: el7host
  user: root
  gather_facts: False
  tasks:
    - script: scripts/createrepo.sh

    - name: cleanup
      file: path=/tmp/testcheckout state=absent force=yes

    - name: create the clone
      git: repo=/tmp/repo.main dest=/tmp/testcheckout

    - name: create a new branch upstream from the current commit
      shell: cd /tmp/clones/repo.main ; git checkout -b BRANCH1 ; git push origin BRANCH1

    - name: checkout the branch in the existing clone
      git: repo=/tmp/repo.main dest=/tmp/testcheckout version=BRANCH1

    - name: check what branch the checkout is on
      shell: cd /tmp/testcheckout ; git branch | head -n1
      register: branch

    - debug: var=branch.stdout

    - assert:
        that:
            - "branch.stdout == '* BRANCH1'"
```

createrepo.sh ...

```
[jtanner@fedmac AMC-612]$ cat scripts/createrepo.sh 
#!/bin/bash

# fix git config
git config --global user.email "root@localhost"
git config --global user.name "rooty mcrootface"

CLONEDIR=/tmp/clones
BASEDIR=/tmp
REPOMAIN=$BASEDIR/repo.main
SUBMODULES="submod1 submod2 submod3 submod4 submod5 submod6"

rm -rf $CLONEDIR
mkdir -p $CLONEDIR
rm -rf $REPOMAIN
mkdir -p $REPOMAIN

# create the primary repo
cd $REPOMAIN
git init --bare

# add submodules to the main
cd $CLONEDIR
git clone $REPOMAIN
cd repo.main
touch main.file 
git add main.file
git commit -m 'INITIAL COMMIT'
git push origin master
```

The run script ...

```
[jtanner@fedmac AMC-612]$ cat test.sh 
#!/bin/bash
VERSION=$(ansible --version | head -n1 | awk '{print $2}')
ansible-playbook -vvvv -i inventory site.yml
RC=$?
exit $RC
```

Results ...

```
v1.7.0 ; 2
v1.7.1 ; 2
v1.7.2 ; 2
v1.8.0 ; 2
v1.8.1 ; 2
v1.8.2 ; 2
v1.8.3 ; 2
v1.8.4 ; 2
v1.9.0-0.1.rc1 ; 3
v1.9.0-0.2.rc2 ; 2
v1.9.0-1 ; 2
v1.9.0-2 ; 2
v1.9.0.1-1 ; 2
v1.9.1-0.1.rc1 ; 2
v1.9.1-0.2.rc2 ; 2
v1.9.1-0.3.rc3 ; 2
v1.9.1-0.4.rc4 ; 2
v1.9.1-1 ; 2
v1.9.2-0.1.rc1 ; 2
v1.9.2-0.2.rc2 ; 2
v1.9.2-1 ; 2
v1.9.3-0.1.rc1 ; 2
v1.9.3-0.2.rc2 ; 2
v1.9.3-0.3.rc3 ; 2
v1.9.3-1 ; 2
v1.9.4-0.1.rc1 ; 2
v1.9.4-0.2.rc2 ; 2
v1.9.4-0.3.rc3 ; 2
v1.9.4-1 ; 2
v1.9.5-0.1.rc1 ; 2
v1.9.5-1 ; 2
v1.9.6-0.1.rc1 ; 2
v1.9.6-1 ; 2
v2.0.0-0.6.rc1 ; 2
v2.0.0-0.7.rc2 ; 2
v2.0.0-0.8.rc3 ; 2
v2.0.0-0.9.rc4 ; 2
v2.0.0.0-1 ; 2
v2.0.0.1-1 ; 2
v2.0.0.2-1 ; 2
v2.0.1.0-0.1.rc1 ; 0
v2.0.1.0-0.2.rc2 ; 0
v2.0.1.0-1 ; 0
v2.0.2.0-0.1.rc1 ; 0
v2.0.2.0-0.2.rc2 ; 0
v2.0.2.0-0.3.rc3 ; 0
v2.0.2.0-0.4.rc4 ; 0
v2.0.2.0-1 ; 0
v2.1.0.0-0.1.rc1 ; 0
v2.1.0.0-0.2.rc2 ; 0
v2.1.0.0-0.3.rc3 ; 0
v2.1.0.0-0.4.rc4 ; 0
v2.1.0.0-1 ; 0
v2.1.1.0-0.1.rc1 ; 0
v2.1.1.0-0.2.rc2 ; 0
v2.1.1.0-0.3.rc3 ; 0
v2.1.1.0-0.4.rc4 ; 0
v2.1.1.0-0.5.rc5 ; 0
v2.1.1.0-1 ; 0
devel-bf8c871 ; 0
```

**Comment by jctanner** [comment]
Based on the results above, this testcase was resolved around v2.0.1.0-0.1.rc1 and has continued working since. @xytis if this is still an issue for you, please create a new ticket with full -vvvv output and a modified version of my test scripts above that clarify what detail I misunderstood.
